src/plasma/plasma_manager.cc

```diff
@@ -586,17 +580,13 @@ void PlasmaManagerState_free(PlasmaManagerState *state) {
     ClientConnection_free(manager_conn);
   }
 
-  if (state->fetch_requests != NULL) {
-    FetchRequest *fetch_req, *tmp;
-    HASH_ITER(hh, state->fetch_requests, fetch_req, tmp) {
-      remove_fetch_request(state, fetch_req);
-    }
-  }
-
-  AvailableObject *entry, *tmp_object_entry;
-  HASH_ITER(hh, state->local_available_objects, entry, tmp_object_entry) {
-    HASH_DELETE(hh, state->local_available_objects, entry);
-    free(entry);
+  /* We have to be careful here because remove_fetch_request modifies
+   * state->fetch_requests in place. */
+  auto it = state->fetch_requests.begin();
+  while (it != state->fetch_requests.end()) {
+    auto next_it = std::next(it, 1);
+    remove_fetch_request(state, it->second);
+    it = next_it;
```

Comment:
This may be unnecessary, it seems like things were working when we just did `it++` after calling `remove_fetch_request`. Note that `remove_fetch_request` removes the value pointed to by `it` from `state->fetch_requests`.

However, `it` is invalidated by the call to `remove_fetch_request`, so this seems safer.

Reply:
And note that we do need to do the cleanup here otherwise valgrind will complain.

Reply:
cc @mehrdadn in case there is an obvious best practice here.

Reply:
Might be simpler to iterate with:
```
while (it != state->fetch_requests.end()) {
  remove_fetch_request(state, it->second);
  it = state->fetch_requests.begin();
}
```

Reply:
This looks a little nicer to me (especially since it should also work with other containers). Want to make that change?

Reply:
I'm actually not sure if removing elements from an `unordered_map` is allowed to change the iterator order or not. If it doesn't, then you would not even go through all the elements. And if it does, chances are someone may get confused about the same thing, so you may want to avoid this for that reason. :)

I think you could just do the following, though for cases where the above isn't an issue, the most "C++" way would be to return an iterator that points past the erased element from `remove_fetch_request`:

```
while (!state->fetch_requests.empty()) {
	remove_fetch_request(state->fetch_requests.begin()->second);
}
```

Reply:
Oops, edited/fixed a typo.^

Reply:
The "best practice" is to just have the container own the value whenever you can, meaning you'd have
```
std::unordered_map<ObjectID, FetchRequest, UniqueIDHasher>
```
or
```
std::unordered_map<ObjectID, std::unique_ptr<FetchRequest>, UniqueIDHasher>
```
as appropriate.

Reply:
Yeah, the right thing to do is to give `fetch_requests` ownership over the `FetchRequest` objects. I'll take a pass at this soon.